tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -18,85 +18,289 @@ limitations under the License.
 };
 
-// Gets the llvm intrinsic ids on different platforms (NVPTX, AMDGPU)
-// corresponding to the give TargetIntrinsicID.
-struct TargetIntrinsics GetIntrinsic(TargetIntrinsicID intrin) {
-  switch (intrin) {
-    case TargetIntrinsicID::kShflDownF32: {
-      return {llvm::Intrinsic::nvvm_shfl_sync_down_f32,
-              llvm::Intrinsic::not_intrinsic};
+// Wrapper structure for  device functionc information for NVPTX/AMDGPU.
+struct TargetFunctionInfo {
+  TargetFunctionInfo()
+      : callee_name(""),
+        input_types({PRIMITIVE_TYPE_INVALID}),
+        result_type(PRIMITIVE_TYPE_INVALID),
+        use_bitcast(false) {}
+  TargetFunctionInfo(const string callee_name_)
+      : callee_name(callee_name_),
+        input_types({PRIMITIVE_TYPE_INVALID}),
+        result_type(PRIMITIVE_TYPE_INVALID),
+        use_bitcast(false) {}
+  TargetFunctionInfo(const string callee_name_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_, bool use_bitcast_)
+      : callee_name(callee_name_),
+        input_types(input_types_),
+        result_type(output_type_),
+        use_bitcast(use_bitcast_) {}
+  // Device function name.
+  const string callee_name;
+  // Inpute types accespted by the device function.
+  absl::Span<const PrimitiveType> input_types;
+  // Result type of the device function.
+  PrimitiveType result_type;
+  // Use bitcast to generate casts if the desired signature at the call site
+  // does not match the signature of the device function.
+  bool use_bitcast;
+} target_function_info;
+
+// Wrapper structure to carry either information about the intrinsic
+// or device function for NVPTX/AMDGPU.
+struct TargetInfo {
+  struct TargetIntrinsicInfo target_intrinsic_info;
+  struct TargetFunctionInfo target_function_info;
+  TargetInfo(struct TargetIntrinsicInfo x, struct TargetFunctionInfo y,
+             bool intrinsic)
+      : target_intrinsic_info(x),
+        target_function_info(y),
+        has_intrinsic(intrinsic) {}
+  TargetInfo(struct TargetIntrinsicInfo x, struct TargetFunctionInfo y)
+      : target_intrinsic_info(x), target_function_info(y) {}
+  absl::optional<bool> has_intrinsic;
+};
+
+// Populates the function information for different platforms (NVPTX, AMDGPU)
+// corresponding to the given TargetFunctionID.
+struct TargetInfo GetTargetInfo(TargetFunctionID function_id,
+                                llvm::Triple& target_triple) {
+  TargetFunctionInfo default_nvptx_function_info, default_amdgpu_function_info;
+  TargetIntrinsicInfo default_nvptx_intrinsic_info, default_amdgpu_intrinsic_info;
+
+  if (!(target_triple.getArch() == llvm::Triple::nvptx ||
+        target_triple.getArch() == llvm::Triple::nvptx64 ||
+        target_triple.getArch() == llvm::Triple::amdgcn)) {
+    LOG(FATAL) << "Invalid triple " << target_triple.str();
+  }
+
+  switch (function_id) {
+    case TargetFunctionID::kShflDownF32:{ 
```

Comment:
@sumesh13 do you feel it be a good idea to break this file into multiple ones? for example, NVPTX-specific ones be in `nvptx_target_util.cc` and AMDGPU-specific ones be in `amdgpu_target_util.cc`? so each path can grow on its own and don't necessarily affect the other.

Reply:
I see how it might be useful.  There was a concern that came up during my upstream review that both structures would need to be updated consistently and so  it would be good to keep it together. Otherwise, its easier to have it separate.

Reply:
got it. thanks for the clarification